### DISCUSSION TITLE
WISH: Control 'repos' order in installation instructions

### DISCUSSION THREAD

**Discussion by HenrikBengtsson** [r-universe-org/help#375] [Ideas]
I'm prototyping [CRANhaven](https://cranhaven.r-universe.dev), which serves as a fallback/backup repository for CRAN.  The auto-generates installation instructions for such packages is:

```r
install.packages('somepkg', repos = c('https://cranhaven.r-universe.dev', 'https://cloud.r-project.org'))
```

However, by design, I want the instructions to use CRANhaven as a last resort, not be the first one. So, something like:

```r
install.packages('somepkg', repos = c('https://cloud.r-project.org', 'https://cranhaven.r-universe.dev'))
```

It would be handy if we could configure/control these installation instructions.

**Comment by dmurdoch** [comment]
I agree.  For most of my packages, the main branch (chosen by default on R-universe) is not a release version, it's a development version.  I've been told I can configure R-universe to choose a release instead, but I like having the easy availability of the devel version, it's just that I don't like the instructions that don't mention that's what you're getting.

So I'd like to customize the installation instructions.  Instead of showing
```
# Install rgl in R:
install.packages('rgl', repos = c('https://dmurdoch.r-universe.dev', 'https://cloud.r-project.org'))
```
I'd choose 
```
# Install the development version of rgl in R:
install.packages('rgl', repos = 'https://dmurdoch.r-universe.dev')

# Install the latest release of rgl:
install.packages('rgl', repos = 'https://cloud.r-project.org')
```

**Comment by jeroen** [comment]
@HenrikBengtsson have you tested your suggestion to work? Afaik, the ordering of repos does not matter much as R will always install the highest version it can find.

**Reply by HenrikBengtsson** [reply]
You're right. I saw a comment about this on R-help/R-devel some weeks ago, but forgot about this feature request.

Closing, since there's nothing you can do here.

**Reply by mojaveazure** [reply]
> Afaik, the ordering of repos does not matter much as R will always install the highest version it can find

There is one exception to this rule: one can pass a `filters` argument to [`install.packages()`/`available.packages()`](https://stat.ethz.ch/R-manual/R-devel/library/utils/html/available.packages.html) that controls how R picks which packages to install. One available filter is `"CRAN"` that tells R to install the highest version of a package from a CRAN repository (named `"CRAN"` in `getOption("repos")`), even if a higher version exists on an alternate repository

For example, with digest, the current [CRAN](https://cran.r-project.org/package=digest) version is 0.6.35 while the current version on [Dirk's R-universe](https://eddelbuettel.r-universe.dev/digest) is 0.6.35.2; as a test (using `rocker/r-ver:latest`)
```R
> getOption('repos')
                                         CRAN
"https://p3m.dev/cran/__linux__/jammy/latest"
> (cran <- available.packages()['digest', 'Version']) # CRAN version of digest
[1] "0.6.35"
> (universe <- available.packages(repos = 'https://eddelbuettel.r-universe.dev/')['digest', 'Version']) # R-universe version of digest
[1] "0.6.35.2"
> cran < universe # check that R-universe has a higher version of digest than CRAN
[1] TRUE
> # basic install.packages(); highest version wins
> install.packages('digest', repos = c('https://eddelbuettel.r-universe.dev/', getOption('repos')), quiet = TRUE)
> packageVersion("digest")
[1] ‘0.6.35.2’
> packageVersion("digest") == universe
[1] TRUE
> # set the filter to use CRAN repos over alternate repos
> options(available_packages_filters = c('R_version', 'OS_type', 'subarch', 'CRAN', 'duplicates'))
> install.packages('digest', repos = c('https://eddelbuettel.r-universe.dev/', getOption('repos')), quiet = TRUE)
> packageVersion("digest")
[1] ‘0.6.35’
> packageVersion("digest") == universe # we have the CRAN version, despite R-universe being passed an higher
[1] FALSE
> # if passing a repos vector without a named CRAN repo (R-universe's default), the CRAN filter does nothing
> install.packages('digest', repos = c('https://eddelbuettel.r-universe.dev/', 'https://cloud.r-project.org'), quiet = TRUE)
> packageVersion("digest")
[1] ‘0.6.35.2’
> packageVersion("digest") == universe
[1] TRUE
```
It's not the simplest solution as it requires
- users to have a default CRAN repository set
- users to alter the default package filters
- R-universe to recommend using `getOption("repos")` over `"https://cloud.r-project.org"`

But having R-universe as a "last resort" should work with the `"CRAN"` filter and a named CRAN repository
